### PR TITLE
seccomp: Improve UX when policy is generated at pod termination

### PR DESCRIPTION
**Pull request by blanquicet**
# Improve UX when policy is generated at pod termination

Currently, when the policy is generated at pod termination, the gadget is reporting an `operationWarning` when it is stopped because there is no pod that matches the requested filter. However, it already generated the policy thus the message does not make sense. Example:

```
$ ./kubectl-gadget seccomp-advisor start -m seccomp-profile --seccomp-profile-name seccomp-demo/hello-profile -n seccomp-demo -p hello-python
DT8HSvzpYJvw3tAp

# Create workload and interact with it

# Terminate workload
$ kubectl delete pod/hello-python -n seccomp-demo
pod "hello-python" deleted

# Check the profile
$ kubectl get seccompprofile -A
NAMESPACE    NAME                   STATUS       AGE
gadget       seccomp-mcnw4-pnpdz    Installed    10s

# Stop monitoring
$ ./kubectl-gadget seccomp-advisor stop DT8HSvzpYJvw3tAp
Failed to run the gadget on all nodes: Pod seccomp-demo/hello-python not found
Error: Failed to run the gadget on all nodes: None of them succeeded
```

## How to use

Try to generate on-demand and at-termination policies and ensure the warning message `Failed to run the gadget on all nodes: Pod my-pod not found` is displayed only in the case the policy is not really generated.

## Testing done

I tested all the uses cases that came to my mind. I tested it on top of https://github.com/kinvolk/inspektor-gadget/pull/421 but it should have no impact.

### Use-case 1 (OutputMode=ExternalResource)
**Description**: The pod is terminated to trigger the policy generation.
**Result**: There is no warning message when the gadget is stopped.

```
$ ./kubectl-gadget seccomp-advisor start -m seccomp-profile --profile-prefix-name hello-profile -n seccomp-demo -p hello-python
L7025nNzD4c4qXel

$ kubectl delete pod/hello-python -n seccomp-demo
pod "hello-python" deleted

# The policy is generated
$ kubectl get seccompprofile -A
NAMESPACE                    NAME                  STATUS      AGE
gadget                       hello-profile-q49wf   Installed   14s

$ ./kubectl-gadget seccomp-advisor stop L7025nNzD4c4qXel
Successfully created seccomp profile
```

### Use-case 2 (OutputMode=ExternalResource)
**Description**: On-demand request to generate the policy but pod does NOT exist. 
**Result**: The warning message is correctly displayed when the gadget is stopped.
 
```
$ ./kubectl-gadget seccomp-advisor start -m seccomp-profile --profile-prefix-name hello-profile -n seccomp-demo -p hello-python
ORQXCfSuVNJ3HhNU

$ ./kubectl-gadget seccomp-advisor stop ORQXCfSuVNJ3HhNU
Failed to run the gadget on all nodes: Pod seccomp-demo/hello-python not found
Error: Failed to run the gadget on all nodes: None of them succeeded
```

### Use-case 3 (OutputMode=ExternalResource)
**Description**: On-demand request to generate the policy and pod does exist.
**Result**: There is no warning message when the gadget is stopped.

```
$ ./kubectl-gadget seccomp-advisor start -m seccomp-profile --profile-prefix-name hello-profile -n seccomp-demo -p hello-python
n1aH1j4B4rVAvfzP

$ ./kubectl-gadget seccomp-advisor stop n1aH1j4B4rVAvfzP
Successfully created seccomp profile

# The policy is generated
$ kubectl get seccompprofile -A
NAMESPACE                    NAME                  STATUS      AGE
gadget                       hello-profile-vswhw   Installed   5s
```

### Use-case 4 (OutputMode=Status)
**Description**: The pod is terminated to trigger the policy generation. Consider it is not supported for OutputMode=Status.
**Result**: The warning message is correctly displayed when the gadget is stopped.

```
$ ./kubectl-gadget seccomp-advisor start -n seccomp-demo -p hello-python
nhVJ7WQcyFyTCziK

$ kubectl delete pod/hello-python -n seccomp-demo
pod "hello-python" deleted

# Any policy is generated. Not supported for OutputMode=Status.

$ ./kubectl-gadget seccomp-advisor stop nhVJ7WQcyFyTCziK
Failed to run the gadget on all nodes: Pod seccomp-demo/hello-python not found
Error: Failed to run the gadget on all nodes: None of them succeeded
```

### Use-case 5 (OutputMode=Status)
**Description**: On-demand request to generate the policy but pod does NOT exist. 
**Result**: The warning message is correctly displayed when the gadget is stopped.

```
$ ./kubectl-gadget seccomp-advisor start -n seccomp-demo -p hello-python
IZZYlgLdJvk8FqNQ

$ ./kubectl-gadget seccomp-advisor stop IZZYlgLdJvk8FqNQ
Failed to run the gadget on all nodes: Pod seccomp-demo/hello-python not found
Error: Failed to run the gadget on all nodes: None of them succeeded
```

### Use-case 6 (OutputMode=Status)
**Description**: On-demand request to generate the policy and pod does exist.
**Result**: There is no warning message when the gadget is stopped.

```
$ ./kubectl-gadget seccomp-advisor start -n seccomp-demo -p hello-python
LDnBqMofyWym671U

$ ./kubectl-gadget seccomp-advisor stop LDnBqMofyWym671U
{
  "defaultAction": "SCMP_ACT_ERRNO",
  "architectures": [
    "SCMP_ARCH_X86_64",
    "SCMP_ARCH_X86",
    "SCMP_ARCH_X32"
  ],
  "syscalls": [
    {
      "names": [
        "epoll_wait",
        "poll",
        "wait4"
      ],
      "action": "SCMP_ACT_ALLOW"
    }
  ]
}
```

### CODE REVIEW

**Comment by blanquicet** [comment]
As an alternative solution to this issue, I was thinking if it is correct the fact that we are considering that a given trace did not succeed if it carries an `OperationWarning` 🤔 :
 
https://github.com/kinvolk/inspektor-gadget/blob/38cafaf070f917015450b17a515cdb47e1786357/cmd/kubectl-gadget/utils/trace.go#L425-L434

**Comment by mauriciovasquezbernal** [comment]
I haven't gone through all the details but I'm wondering why there is not a 1-1 mapping between the commands in the CLI and the operations in the trace CR. I see the trace supports `start`, `generate` and `stop` and the cli has `start` and `stop`. Is the issue your mentioning here related to the fact that the CLI is combining the generate + stop operations in the stop command?

**Comment by blanquicet** [comment]
> Is the issue your mentioning here related to the fact that the CLI is combining the generate + stop operations in the stop command?
>
Partially. I mean, even if we added the `generate` operation to the CLI, the same issue would happen if the user does: `seccomp-advisor start`, terminate the workload and then `seccomp-advisor generate`. However, I would expect that the user will execute `seccomp-advisor stop` instead of `seccomp-advisor generate` knowing that the policy was already generated when the workload terminated, it is something we could assume? I am not completely sure but if so, the issue would not happen.

> I haven't gone through all the details but I'm wondering why there is not a 1-1 mapping between the commands in the CLI and the operations in the trace CR. I see the trace supports start, generate and stop and the cli has start and stop.
>
We are discussing making it 1:1 but by removing the `generate` operation from the trace CR as an effort to uniform all the gadgets (https://github.com/kinvolk/inspektor-gadget/issues/367). Consider that seccomp is the only trace CR that does not follow the `start/stop` approach.

**Comment by eiffel-fl** [comment]
Hi.

I also think in the future the `Generate()` and `Stop()` of seccompadvisor gadget will be merged, so I do not worry about them for the moment.

Regarding this PR, it is cool! I quickly check the last commit and it is OK for me.
But let's wait for #421 to be merged first as you built this one on top of it.

Best regards.

File: pkg/gadgets/seccomp/gadget.go [hunk 1]
```diff
@@ -41,6 +41,11 @@ type Trace struct {
 	client   client.Client
 
 	started bool
+
+	// policyAtTermination is used to know if there was a policy generated
+	// at pod termination so that the Generate() operation does not have
+	// to notify that it did not find a pod that matches the filter.
+	policyAtTermination bool
```
Comment (mauriciovasquezbernal) [inline pkg/gadgets/seccomp/gadget.go:48]: What about calling it `policyGenerated`? Or something that indicates what this variable is for: To indicate that a policy was generated and we don't need to throw and error.

**Review by mauriciovasquezbernal — APPROVED** [review]
Just a comment on the variable name. Otherwise LGTM!

File: pkg/gadgets/seccomp/gadget.go (hunk 1 above)
Comment (blanquicet) [reply pkg/gadgets/seccomp/gadget.go:48]: Absolutely! I prefer `policyGenerated`.

**Comment by blanquicet** [comment]
PR rebased on top of main.

**Pull request merged**